lib/hammer_cli/defaults.rb

```diff
@@ -0,0 +1,207 @@
+    end
+
+    def bad_input
+      print_message(_("You must specify value or a plugin name, cant specify both."))
+    end
+
+    def variable_not_found
+      print_message(_("Couldn't find the requested param in %s.") % context[:defaults].path)
+    end
+
+    def list_all_defaults_message
+      unless context[:defaults].defaults_settings.nil?
+        print_message("--------------------")
+        print_message("Hammer defaults list")
+        print_message("--------------------")
+        context[:defaults].defaults_settings.each do |key,val|
+          if val[:from_server]
+            print_message(key.to_s + " : " +  _("(provided by %{plugin})") % {:plugin => val[:provider].to_s.split(':').first.gsub("HammerCLI", '')})
+          else
+            print_message(key.to_s + " : " + val[:value].to_s)
+          end
+        end
+      else
+        print_message(_("No defaults file was found"))
+      end
+    end
+
+    autoload_subcommands
+  end
+
+
+  class Defaults
+    DEFAULT_FILE = "#{Dir.home}/.hammer/cli.modules.d/defaults.yml"
+
+    class DefaultsError < StandardError; end
+    class DefaultsPathError < DefaultsError; end
+
+    attr_reader :defaults_settings
+
+    def initialize(settings, file_path = nil)
+      @defaults_settings = settings
+      @path = file_path || DEFAULT_FILE
+    end
+
+    def register_provider(provider)
+      providers[provider.plugin_name.to_s] = provider
+    end
+
+    def providers
+      @providers ||= {}
+    end
+
+    def delete_default_from_conf(param)
+      conf_file = YAML.load_file(path)
+      conf_file[:defaults].delete(param)
+      write_to_file conf_file
+      conf_file
+    end
+
+    def add_defaults_to_conf(default_options, provider)
+      create_default_file if defaults_settings.nil?
+      defaults = YAML.load_file(path)
+      defaults[:defaults] ||= {}
+      default_options.each do |key, value|
+        key = key.to_sym
+        defaults[:defaults][key] = value ? {:value => value, :from_server => false} : {:from_server => true, :provider => provider}
+      end
+      write_to_file defaults
+      defaults
+    end
+
+    def get_defaults(option)
+      unless defaults_settings.nil? || defaults_settings[option.to_sym].nil?
+        if defaults_settings[option.to_sym][:from_server]
+          providers[defaults_settings[option.to_sym][:provider]].get_defaults(option.to_sym)
+        else
+          defaults_settings[option.to_sym][:value]
+        end
+      end
+    end
+
+    def write_to_file(defaults)
+      File.open(path,'w') do |h|
+        h.write defaults.to_yaml
+      end
+    end
+
+    protected
+
+    attr_reader :path
+
+    def create_default_file
+      if Dir.exist?(File.dirname(@path))
+        new_file = File.new(path, "w")
+        new_file.write ":defaults:"
+        new_file.close
+      else
+        raise DefaultsPathError.new(_("Couldn't create %s please create the path before defaults will be enabled.") % path)
```

Comment:
Sounds "Couldn't create %s please create the path before defaults are enabled."  better?